### ISSUE TITLE
Cannot easily override GC settings via configuration

### ISSUE THREAD

**Issue by j-baker** [palantir/gradle-java-distribution#998]
sls-packaging encourages the use of gc defaults, e.g. 'response-time', 'hybrid', 'throughput'. These use cms/shenandoah, g1, throughput collectors in usually sensible ways.

Unfortunately, they are specifiable only once, at service authorship time, and never at configuration time. This makes things tricky. If one e.g. wants to switch from response-time to hybrid via config, one must write things like `-XX:-UseConcMarkSweepGC` to disable the previously added flags. But not all of the flags are actually unsettable, e.g. new gen ratio. This means that we poorly support a few use cases:

1. Trying out a different garbage collector. Ideally we'd be able to test for an extended period on a single installation, but right now that means tagging an RC release with the new profile and holding back all other upgrades for that deployment.
2. Setups where a single installation has a very different workload for all others. For example, we have a setup where most deployments have a 16GB heap, but one has a 64GB heap and another has a 256GB heap. In such a case, many of the default settings (e.g. new size ratio) become incorrect. This is clearly an antipattern, but we've multiple times found that fixing the issues that cause you to need a 256GB heap is a process that takes months if not years. We don't end up with a 256GB heap via laziness, we end up with that because it's the local maximum.

I think the intent of this was to remove as much from configuration as possible, but in practice it adds friction to testing out GC changes and instead has led in at least one case to service binary monkey patching (e.g. manually modifying launcher-static.yml (see PDS-126032 internally)) (which seems worse than the alternative). It would be great if we could specify one of the gc profiles via launcher custom.

I think there are some impediments in the implementation; right now I think the gc profiles are implemented purely at the gradle level. If the go-java-launcher were to have a little more 'sympathy' for how we define configuration internally then we'd have less issue. As one brief proposal, you could imagine launcher static containing maps of options vs simply a list

```yaml
profile-definitions:
  # populated automatically by gradle plugin
  gc-hybrid:
    - "-XX:+UseG1GC"
  gc-response-time:
   - "-XX:+UseConcMarkSweepGC"
  # added by user
  experimental-disable-biased-locking:
   - "-XX:-UseBiasedLocking"
profiles:
  gc-hybrid: true
```

and then launcher custom could contain something like:

```yaml
configType: java
configVersion: 1 # this is fully backwards compatible
profiles:
  gc-hybrid: false
  gc-response-time: true # hybrid didn't work very well
  experimental-disable-biased-locking: true
jvmOpts:
  - deprecated
  - please
  - use
  - profiles
  - this
  - just ensures that we can actually roll this out
```

where by using a map, this is 1) more sympathetic to our deployment infrastructure and 2) enables disabling of statically enabled profiles like gc where necessary.

This obviously sits against our desire to make as few things dynamically overridable as possible. My belief is that sls-packaging exists at too low a level for the opinion to be enforceable in a way that doesn't cause frustration.

In terms of the impact to this, I think I've seen at least 5-10 cases where this has caused friction. I can detail internally. Most relate to trying out new things (e.g. tagging RCs to try out Shenandoah) but a number relate to production use cases.

**Comment by ypan0411** [comment]
@iamdanfox @ferozco Can I get an eye on this issue? I would like to know if it would be possible to prioritize this work from your side? If not, I can try to do external contribute, which might need your help as well. Thanks!

**Comment by fawind** [comment]
Chatted with @ferozco about this. We plan on adding an additional GC profile that does not set any flags. This option together with setting custom flags should allow you to test out GC changes more freely without having the issue of unsetting flags.

**Comment by carterkozak** [comment]
Closing this, it was resolved by #1015
